File: app/src/main/java/me/zhanghai/android/douya/network/api/ApiContract.java
/*
 * Copyright (c) 2015 Zhang Hai <Dreaming.in.Code.ZH@Gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.douya.network.api;

import android.os.Build;

import java.util.Arrays;
import java.util.List;

import me.zhanghai.android.douya.network.Http;

public interface ApiContract {

    interface Request {

        interface Authentication {

            interface BaseUrls {
                String API_V2 = "https://www.douban.com/";
                String FRODO = "https://frodo.douban.com/";
            }
            String URL = "service/auth2/token";

            String ACCEPT_CHARSET = Http.Charsets.UTF8;
            interface RedirectUris {
                String API_V2 = "http://shuo.douban.com/!service/android";
                String FRODO = "frodo://app/oauth/callback/";
            }
            interface GrantTypes {
                String PASSWORD = "password";
                String REFRESH_TOKEN = "refresh_token";
            }
        }

        interface Base {
            int MAX_NUM_RETRIES = 2;
        }

        interface Frodo {

            String BASE_URL = "https://frodo.douban.com/api/v2/";

            // API protocol version is derived from user agent string.
            String USER_AGENT = "api-client/1 com.douban.frodo/6.0.1(138) Android/" +
                    Build.VERSION.SDK_INT+ " product/" + Build.PRODUCT + " vendor/" +
                    // Sorry Frodo, but we don't want to hold ACCESS_NETWORK_STATE.
                    Build.MANUFACTURER + " model/" + Build.MODEL + "  rom/android  network/wifi";

            String API_KEY = "apikey";
            String CHANNEL = "channel";
            interface Channels {
                String DOUBAN = "Douban";
            }
            String UDID = "udid";
            String OS_ROM = "os_rom";
            interface OsRoms {
                String ANDROID = "android";
            }

            List<String> SIGNATURE_HOSTS = Arrays.asList("frodo.douban.com", "api.douban.com");
            String SIG = "_sig";
            String TS = "_ts";
        }

        interface ApiV2 {

            String BASE_URL = "https://api.douban.com/v2/";

            String USER_AGENT = "api-client/2.0 com.douban.shuo/2.2.7(123) Android/" +
                    Build.VERSION.SDK_INT + " " + Build.PRODUCT + " " + Build.MANUFACTURER + " " +
                    Build.MODEL;

            interface Base {
                String API_KEY = "apikey";
                String UDID = "udid";
            }

            interface LifeStream {
                String VERSION = "version";
                interface Versions {
                    int TWO = 2;
                }
            }
        }
    }

    interface Response {

        interface Error {
            String CODE = "code";
            interface Codes {
                interface Custom {
                    int INVALID_ERROR_RESPONSE = -1;
                }
                interface Base {
                    int INVALID_REQUEST_997 = 997;
                    int UNKNOWN_V2_ERROR = 999;
                    int NEED_PERMISSION = 1000;
                    int URI_NOT_FOUND = 1001;
                    int MISSING_ARGS = 1002;
                    int IMAGE_TOO_LARGE = 1003;
                    int HAS_BAN_WORD = 1004;
                    int INPUT_TOO_SHORT = 1005;
                    int TARGET_NOT_FOUND = 1006;
                    int NEED_CAPTCHA = 1007;
                    int IMAGE_UNKNOWN = 1008;
                    int IMAGE_WRONG_FORMAT = 1009;
                    int IMAGE_WRONG_CK = 1010;
                    int IMAGE_CK_EXPIRED = 1011;
                    int TITLE_MISSING = 1012;
                    int DESC_MISSING = 1013;
                }
                interface Token {
                    int INVALID_REQUEST_SCHEME = 100;
                    int INVALID_REQUEST_METHOD = 101;
                    int ACCESS_TOKEN_IS_MISSING = 102;
                    int INVALID_ACCESS_TOKEN = 103;
                    int INVALID_APIKEY = 104;
                    int APIKEY_IS_BLOCKED = 105;
                    int ACCESS_TOKEN_HAS_EXPIRED = 106;
                    int INVALID_REQUEST_URI = 107;
                    int INVALID_CREDENCIAL1 = 108;
                    int INVALID_CREDENCIAL2 = 109;
                    int NOT_TRIAL_USER = 110;
                    int RATE_LIMIT_EXCEEDED1 = 111;
                    int RATE_LIMIT_EXCEEDED2 = 112;
                    int REQUIRED_PARAMETER_IS_MISSING = 113;
                    int UNSUPPORTED_GRANT_TYPE = 114;
                    int UNSUPPORTED_RESPONSE_TYPE = 115;
                    int CLIENT_SECRET_MISMATCH = 116;
                    int REDIRECT_URI_MISMATCH = 117;
                    int INVALID_AUTHORIZATION_CODE = 118;
                    int INVALID_REFRESH_TOKEN = 119;
                    int USERNAME_PASSWORD_MISMATCH = 120;
                    int INVALID_USER = 121;
                    int USER_HAS_BLOCKED = 122;
                    int ACCESS_TOKEN_HAS_EXPIRED_SINCE_PASSWORD_CHANGED = 123;
                    int ACCESS_TOKEN_HAS_NOT_EXPIRED = 124;
                    int INVALID_REQUEST_SCOPE = 125;
                    int INVALID_REQUEST_SOURCE = 126;
                    int THIRDPARTY_LOGIN_AUTH_FAILED = 127;
                    int USER_LOCKED = 128;
                }
                interface Followship {
                    int ALREADY_FOLLOWED = 10003;
                    int NOT_FOLLOWED_YET = 10005;
                }
                interface Broadcast {
                    int NOT_FOUND = 1212;
                }
                interface RebroadcastBroadcast {
                    int REBROADCASTED_BROADCAST_DELETED = 1265;
                }
            }
            String MSG = "msg";
            String REQUEST = "request";
            String LOCALIZED_MESSAGE = "localized_message";
        }
    }
}

File: app/src/main/java/me/zhanghai/android/douya/api/info/ApiContract.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.douya.api.info

import android.os.Build
import android.util.Base64
import me.zhanghai.android.douya.R
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

object ApiContract {
    object Api {
        const val MAX_AUTH_RETRIES = 2

        const val BASE_URL = "https://frodo.douban.com/api/v2/"

        // API protocol version is derived from user agent string.
        val USER_AGENT = "api-client/1 com.douban.frodo/6.28.1(174) Android/" +
            Build.VERSION.SDK_INT + " product/" + Build.PRODUCT + " vendor/" + Build.MANUFACTURER +
            // Sorry Frodo, but we don't want to request ACCESS_NETWORK_STATE.
            " model/" + Build.MODEL + "  rom/android  network/wifi  platform/mobile"

        const val API_KEY = "apikey"

        const val CHANNEL = "channel"
        object Channels {
            const val DOUBAN = "Douban"
        }

        const val UDID = "udid"

        const val OS_ROM = "os_rom"
        object OsRoms {
            const val ANDROID = "android"
        }

        val SIGNATURE_HOSTS = listOf("frodo.douban.com", "api.douban.com")
        const val SIG = "_sig"
        const val TS = "_ts"
    }

    object Credential {
        val KEY: String

        val SECRET: String

        init {
            val cipher = Cipher.getInstance("AES/CBC/NoPadding")
            val key = SecretKeySpec("MIICUjCCAbsCBEty".toByteArray(Charsets.UTF_8), "AES")
            val parameters = IvParameterSpec("DOUBANFRODOAPPIV".toByteArray(Charsets.UTF_8))
            cipher.init(Cipher.DECRYPT_MODE, key, parameters)
            val keyInput = Base64.decode(
                "74CwfJd4+7LYgFhXi1cx0IQC35UQqYVFycCE+EVyw1E=", Base64.DEFAULT
            )
            KEY = cipher.doFinal(keyInput).toString(Charsets.UTF_8)
            val secretInput = Base64.decode("bHUvfbiVZUmm2sQRKwiAcw==", Base64.DEFAULT)
            SECRET = cipher.doFinal(secretInput).toString(Charsets.UTF_8)
        }
    }

    object Authentication {
        const val BASE_URL = "https://frodo.douban.com/"

        const val URL = "service/auth2/token"

        const val REDIRECT_URI = "frodo://app/oauth/callback/"

        object GrantTypes {
            const val PASSWORD = "password"
            const val REFRESH_TOKEN = "refresh_token"
        }
    }

    object Error {
        object Codes {
            const val INVALID_ACCESS_TOKEN = 103
            const val ACCESS_TOKEN_HAS_EXPIRED = 106
            const val USERNAME_PASSWORD_MISMATCH = 120
            const val INVALID_USER = 121
            const val USER_IS_BLOCKED = 122
            const val ACCESS_TOKEN_HAS_EXPIRED_SINCE_PASSWORD_CHANGED = 123
            const val USER_IS_LOCKED = 128
        }

        val MESSAGES = mapOf(
            Codes.INVALID_ACCESS_TOKEN to R.string.api_error_invalid_access_token,
            Codes.ACCESS_TOKEN_HAS_EXPIRED to R.string.api_error_access_token_has_expired,
            Codes.USERNAME_PASSWORD_MISMATCH to R.string.api_error_username_password_mismatch,
            Codes.INVALID_USER to R.string.api_error_invalid_user,
            Codes.USER_IS_BLOCKED to R.string.api_error_user_is_blocked,
            Codes.ACCESS_TOKEN_HAS_EXPIRED_SINCE_PASSWORD_CHANGED
                to R.string.api_error_access_token_has_expired_since_password_changed,
            Codes.USER_IS_LOCKED to R.string.api_error_user_is_locked
        )
    }
}
